replica/table.cc

```diff
@@ -652,6 +659,30 @@ future<> table::parallel_foreach_compaction_group(std::function<future<>(compact
     });
 }
 
+future<sstables::sstable_list> table::take_storage_snapshot(dht::token_range tr) {
+    sstables::sstable_list ret;
+
+    for (auto cg_id : compaction_groups_for_token_range(tr)) {
+        auto& cg = _compaction_groups[cg_id];
+
+        // We don't care about sstables in snapshot being unlinked, as the file
+        // descriptors remain opened until last reference to them are gone.
+        // Also, we should be careful with taking a deletion lock here as a
+        // deadlock might occur due to memtable flush backpressure waiting on
+        // compaction to reduce the backlog.
+
+        co_await cg->flush();
+
+        auto main_ssts = cg->main_sstables()->all();
+        auto maintenance_ssts = cg->maintenance_sstables()->all();
```

Comment:
This is a bit fragile and will breakIf we add another sstable set, e.g. staging sstables.
Maybe post my series to make cache update exception safe we could add a compound set at the compaction group level and use it here to get all sstables.

Reply:
Also, there's a for_each_sstable interface now so you don't need to generate and store a temp copy of `all()`

Reply:
you can use .size() to reserve capacity in `ret` and iterate using for_each_sstable and insert them one at a time.

Reply:
> you can use .size() to reserve capacity in `ret` and iterate using for_each_sstable and insert them one at a time.

if you don't know all size in advance, reserve() can do more harm than good. with unordered_set, the set is being rehashed on every reserve, rather than log2 N times.